L74-java-channel-creds.md

```diff
@@ -0,0 +1,489 @@
+    .build().start();
+```
+
+
+```java
+package io.grpc;
+
+public abstract class ServerCredentials {}
+```
+
+Since servers using TLS require a certificate, the `TlsServerCredentials`
+cannot have the simple no-arg configuration of `TlsChannelCredentials`. It will
+provide convenience constructors for providing a certificate and unencrypted
+key. If any other configuration is necessary, the `Builder` would be used.
+
+```java
+package io.grpc;
+
+public final class TlsServerCredentials extends ServerCredentials {
+  /**
+   * Creates an instance using provided certificate chain and private key.
+   * Generally they should be PEM-encoded and the key is an unencrypted PKCS#8
+   * key (file headers have "BEGIN CERTIFICATE" and "BEGIN PRIVATE KEY").
+   */
+  public TlsServerCredentials(File certChain, File privateKey)
+      throws IOException {...}
+  // Will not auto-close InputStream, unlike existing transportSecurity()
+  public TlsServerCredentials(InputStream certChain, InputStream privateKey)
+      throws IOException {...}
+
+  /** Same as client-side */
+  public Set<Feature> incomprehensible(Set<Feature> understoodFeatures) {...}
+
+  public byte[] getCertificateChain() { return Arrays.copyOf(...); }
+  public byte[] getPrivateKey() { return Arrays.copyOf(...); }
+  public String getPassword() {...}
+
+  /* Other methods can be added to support various Features */
+
+  public enum Feature {}
+
+  public static Builder newBuilder() { return new Builder(); }
+
+  public final static class Builder {
+    public Builder keyManager(File certChain, File privateKey)
+        throws IOException {...}
+    public Builder keyManager(File certChain, File privateKey, String password)
+        throws IOException {...}
+    public Builder keyManager(InputStream certChain, InputStream privateKey)
+        throws IOException {...}
+    public Builder keyManager(
+        InputStream certChain, InputStream privateKey, String password)
+        throws IOException {...}
+
+    // Throws if no certificate was provided
+    public TlsChannelCredentials build() {...}
+  }
+}
+```
+
+## Rationale
+
+A channel credential API is substantially harder in Java than the other
+languages as Java has multiple transports that behave radically differently.
+OkHttp uses blocking I/O with a byte[]-backed Buffer whereas Netty uses
+nonblocking I/O with a reference-counted, ByteBuffer-backed ByteBuf with an
+event loop. Go and C use a code-based API where credentials implement an
+interface and use a single I/O model. It is not practical to have one
+code-based API to support both OkHttp and Netty, as it would have a
+considerable amount of complexity in the implementations, be hard to optimize,
+and have a wide public API surface. Instead, Java needs a data-based API, where
+configuration values are communicated and each transport can have its own
+implementation. A data-based API can also be used by InProcess and Binder
+transports.
```

Comment:
Great rationale. A diagram showing the OkHttp and Netty blocks, and blocks for Channel credentials and how the transport blocks (Netty and OkHttp) fetch configs from the channel creds to implement security implied by the config. Just a suggestion.

Reply:
It isn't clear to me what this would look like. Maybe we should discuss "in person"?

Reply:
ok